yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -223,4 +225,43 @@ object YarnSparkHadoopUtil {
     )
   }
 
+  /**
+   * Expand environment variable using Yarn API.
+   * If environment.$$() is implemented, return the result of it.
+   * Otherwise, return the result of environment.$()
+   * Note: $$() is added in Hadoop 2.4.
+   */
+  def expandEnvironment(environment: Environment): String = {
```

Comment:
Also, is this really expanding the variable? It seems like it's just creating a reference to the variable. e.g.,

    expandEnvironment(Environment.PWD)

becomes

    "{{PWD}}"

Not the actual value of `$PWD`, right?